mantidproject/mantid
# Add custom tt mode for PEARL

**Description of work.**

Add a new tt_mode value of "custom" to the PEARL powder diffraction scripts. Previously the PEARL users have been using one of the pre-existing tt_mode values and overwriting the standard grouping file with a custom one temporarily. But this involved some manual messing around and was more error prone. Also the code assumes there are 14 groups at the moment which may not actually be the case.

**Report to:** [Chris Ridley]

**To test:**

Unzip the attached zip file somewhere. Update the C:\PEARL\ paths in test.py and User_experiment.yaml to match where you saved the zip file. Run the test.py file (which contains the following) and check it completes successfully. You will need access to the ISIS data archive:
```
# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *
import matplotlib.pyplot as plt
import numpy as np

from isis_powder.pearl import Pearl

pearl_focus = Pearl(user_name="Danny",
                          output_directory=r"C:\PEARL",   #directory above cycle number 
                          config_file=r"C:\PEARL\User_experiment.yaml",
                          calibration_directory=r"C:\PEARL\calib_critical_files",
                          calibration_mapping_file=r"C:\PEARL\calib_critical_files\pearl.yaml")

pearl_focus.create_vanadium(run_in_cycle=116374, do_absorb_corrections=True, tt_mode='custom',
                  custom_grouping_filename=r"C:\PEARL\DAC_group.cal")

pearl_focus.focus(run_number="116374", file_ext=".s02", perform_attenuation=False, 
                  subtract_empty_instrument=False, tt_mode='custom', 
                  custom_grouping_filename=r"C:\PEARL\DAC_group.cal")
```

[test.zip](https://github.com/mantidproject/mantid/files/6688468/test.zip)

The script uses a custom grouping file that Chris sent me. You should see a set out ~6 output files with the mode "custom" in the name in a "21_1" subfolder beneath the output folder:

![image](https://user-images.githubusercontent.com/56295817/122881440-11c75100-d333-11eb-9843-ab527ad19c3a.png)


Fixes #31779 .





---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
